src/swap/defi/thorchain.js

```diff
@@ -0,0 +1,611 @@
+      }
+      const { address: thorAddress, gas_rate: gasRate } = inAddressObject
+      clog(`${fromMainnetCode}.${fromCurrencyCode} gasRate ${gasRate}`)
+
+      const outAddressObject = inboundAddresses.find(
+        addrObj => addrObj.halted === false && addrObj.chain === toMainnetCode
+      )
+      if (outAddressObject == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+      const { gas_rate: outboundGasRate } = outAddressObject
+      clog(
+        `${toMainnetCode}.${toCurrencyCode} outboundGasRate ${outboundGasRate}`
+      )
+
+      const sourcePool = pools.find(pool => {
+        const [asset] = pool.asset.split('-')
+        return asset === `${fromMainnetCode}.${fromCurrencyCode}`
+      })
+      if (sourcePool == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+      const [sourceAsset] = sourcePool.asset.split('-')
+      clog(`sourceAsset: ${sourceAsset}`)
+
+      const destPool = pools.find(pool => {
+        const [asset] = pool.asset.split('-')
+        return asset === `${toMainnetCode}.${toCurrencyCode}`
+      })
+      if (destPool == null) {
+        throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
+      }
+
+      // Add outbound fee
+      const networkFee = calcNetworkFee(
+        toMainnetCode,
+        toCurrencyCode,
+        outboundGasRate,
+        pools
+      )
+
+      // Per Thorchain specs, user pays 2x the outbound transaction networkfee
+      let feeInDestCurrency = mul(networkFee, '2')
+      clog(`feeInDestCurrency: ${feeInDestCurrency}`)
+
+      const assetInUsd = mul(feeInDestCurrency, destPool.assetPriceUSD)
+      if (lt(assetInUsd, '1')) {
+        feeInDestCurrency = div('1', destPool.assetPriceUSD, DIVIDE_PRECISION)
+        clog(`feeInDestCurrency adjusted to $1 min: ${feeInDestCurrency}`)
+      }
+
+      let calcResponse
+      if (quoteFor === 'from') {
+        calcResponse = await calcSwapFrom(
+          {
+            fromWallet,
+            fromCurrencyCode,
+            toWallet,
+            toCurrencyCode,
+            nativeAmount,
+            sourcePool,
+            destPool,
+            volatilitySpreadFinal,
+            affiliateFee,
+            feeInDestCurrency
+          },
+          clog
+        )
+      } else {
+        throw new Error('TODO: TO quoting')
+      }
+      const { fromNativeAmount, toNativeAmount, limit } = calcResponse
+
+      let customNetworkFee
+      let customNetworkFeeKey
+
+      const customFeeTemplate = (fromWallet.currencyInfo.customFeeTemplate ??
+        [])[0]
+      const fromCurrencyInfo = fromWallet.currencyInfo
+      if (customFeeTemplate?.type === 'nativeAmount') {
+        customNetworkFee = gasRate
+        customNetworkFeeKey = customFeeTemplate.key
+      } else if (fromCurrencyInfo.defaultSettings?.customFeeSettings != null) {
```

Comment:
This code will almost certainly break in the future, since `defaultSettings` is an `any` dumping ground. I see that you are using the "official" way on the lines above, which tells me we probably need to get our currency plugins upgraded to use that instead.

Reply:
I'm all for getting everything upgraded to the correct format, but I didn't want to scope creep this project so I adopted whatever models are currently in use. I won't make any changes to this PR for now.